Require provider-defined function parameter naming

bendbennett opened the pull request:
Terraform core requires that all parameters that are used in the context of a provider-defined function are named. The plugin framework has previously either allowed provider developers to either explicitly supply names for provider-defined function parameters, or to generate default parameter names (e.g., `param1`, `param2`, `varparam`) when explicit names are absent.

This PR removes the generation of default parameter names by the plugin framework and therefore requires that explicit names are supplied for all provider-defined function parameters. This change was made in order to try and reduce any potential ambiguity for practitioners arising from the Terraform CLI outputting default parameter names. For example, prior to the changes in this PR, an error arising from, for example, a missing parameter value, would give rise to an error of the following form:

```shell
│ Error: Not enough function arguments
│ 
│   on resource.tf line 10, in resource "example_resource" "example":
│   10:   configurable_attribute = provider::example::string_len("some-values")
│     ├────────────────
│     │ while calling provider::example::string_len(param1, param2)
│ 
│ Function "provider::example::string_len" expects 2 argument(s). Missing value for "param2".
```   

### Breaking Change

Any provider-defined functions that do not have explicitly defined parameter names will now generate a runtime error of the following form:

```shell
│ Error: Failed to load plugin schemas
│ 
│ Error while loading schemas for plugin components: Failed to obtain provider schema: Could not load the schema for provider registry.terraform.io/bendbennett/playground: failed to
│ retrieve schema from provider "registry.terraform.io/bendbennett/playground": Invalid Function Definition: When validating the function definition, an implementation issue was
│ found. This is always an issue with the provider and should be reported to the provider developers.
│ 
│ Function "string_len" - Parameter at position 0 does not have a name..
```

Errors of this type can be fixed by explicitly naming all provider-defined function parameters. For example:

```go
func (f ExampleFunction) Definition(ctx context.Context, req function.DefinitionRequest, resp *function.DefinitionResponse) {
	resp.Definition = function.Definition{
		Parameters: []function.Parameter{
			function.StringParameter{
				Name: "input", // Name explicitly supplied
			},
		},
		Return: function.StringReturn{},
	}
}
```

References:

- [Error when loading plugin schemas outputs diagnostics suffixed with double period](https://github.com/hashicorp/terraform/issues/34865#top)
- hashicorp/terraform-provider-corner#227

bflad commented on .changes/unreleased/BREAKING CHANGES-20240320-095152.yaml:
```diff
@@ -0,0 +1,5 @@
+kind: BREAKING CHANGES
+body: 'function: All parameters must be explicitly named'
```
```suggestion
body: 'function: All parameters must be explicitly named via the `Name` field'
```

bflad commented on function/definition.go:
```diff
@@ -89,41 +89,46 @@ func (d Definition) Parameter(ctx context.Context, position int) (Parameter, dia
 // logic runs during the GetProviderSchema RPC, or via provider-defined unit
 // testing, and should never include false positives.
-func (d Definition) ValidateImplementation(ctx context.Context) diag.Diagnostics {
+func (d Definition) ValidateImplementation(ctx context.Context, funcName string) diag.Diagnostics {
```
Since this is an exported function, should we switch this to request/response types while we are making a breaking change to the signature and to allow for future-proofing?

bendbennett replied:
Good call. Have added request-response pattern for `ValidateImplementation` along with `DefinitionValidateRequest` and `DefinitionValidateResponse`.

bflad commented on function/parameter.go:
```diff
@@ -7,17 +7,6 @@ import (
 	"github.com/hashicorp/terraform-plugin-framework/attr"
 )
 
-const (
-	// DefaultParameterNamePrefix is the prefix used to default the name of parameters which do not declare
-	// a name. Use this to prevent Terraform errors for missing names. This prefix is used with the parameter
-	// position in a function definition to create a unique name (param1, param2, etc.)
-	DefaultParameterNamePrefix = "param"
-
-	// DefaultVariadicParameterName is the default name given to a variadic parameter that does not declare
-	// a name. Use this to prevent Terraform errors for missing names.
-	DefaultVariadicParameterName = "varparam"
-)
```
For breaking change completeness, we'll need to also call out these constant removals.

bendbennett replied:
Have added a changelog entry for this.

bflad reviewed (approved):
A few small things, otherwise looks good to me 👍

austinvalle reviewed (approved):
Looks awesome, great work Ben!

One note: we'll also want to clean up the documentation to reflect these changes, from a quick glance, some pages I found that should be updated:
- https://developer.hashicorp.com/terraform/plugin/framework/functions/implementation#reading-argument-data (example code doesn't have names)
- https://developer.hashicorp.com/terraform/plugin/framework/functions/documentation#parameters (mentions old defaulting pattern of `param1`, etc.)
  - I wonder if we should give an example of an error message when you have a named parameter? Just to give the provider developer a visual of how it will be used (we mention it's used in docs/errors already)
- https://developer.hashicorp.com/terraform/plugin/framework/functions/parameters/bool (Pretty much all of the type specific parameter docs have an example that doesn't have `Name` defined)

There may be more, but that's what I initially found 👍🏻

bendbennett commented:
> Looks awesome, great work Ben!
> 
> One note: we'll also want to clean up the documentation to reflect these changes, from a quick glance, some pages I found that should be updated:
> 
> * https://developer.hashicorp.com/terraform/plugin/framework/functions/implementation#reading-argument-data (example code doesn't have names)
> * https://developer.hashicorp.com/terraform/plugin/framework/functions/documentation#parameters (mentions old defaulting pattern of `param1`, etc.)
>   
>   * I wonder if we should give an example of an error message when you have a named parameter? Just to give the provider developer a visual of how it will be used (we mention it's used in docs/errors already)
> * https://developer.hashicorp.com/terraform/plugin/framework/functions/parameters/bool (Pretty much all of the type specific parameter docs have an example that doesn't have `Name` defined)
> 
> There may be more, but that's what I initially found 👍🏻

I've done a search through the docs for all examples of parameter usage and added `Name` to each example.

I added an addition subsection for `Parameter Naming` to `terraform/plugin/framework/functions/parameters#parameter-naming`, which includes examples of runtime errors generated if unnamed parameters are present, and errors generated when there is an issue with a specific parameter.

austinvalle commented:
Looks great ❤️ , I updated the dynamic docs PR to follow that pattern 👍🏻 
https://github.com/hashicorp/terraform-plugin-framework/pull/954/commits/c75a5303fa310935b9fda7037a5cad9b1e4c2fae

bflad commented on website/docs/plugin/framework/functions/documentation.mdx:
```diff
@@ -45,11 +45,11 @@ func (f *CidrContainsIpFunction) Definition(ctx context.Context, req function.De
-| `MarkdownDescription` | Documentation about the parameter and its expected values in Markdown format. |
+| Field Name            | Description                                                                                                                                           |
+|-----------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `Name`                | **Required**: Single word or abbreviation of parameter for function signature generation. If name is not provided, a runtime error will be generated. |
```
❤️

bflad commented on website/docs/plugin/framework/functions/implementation.mdx:
```diff
@@ -205,10 +217,14 @@ func (f *ExampleFunction) Definition(ctx context.Context, req function.Definitio
         },
         VariadicParameter: function.StringParameter{
             Name: "variadic_param",
+			// ... other fields ...
```
```suggestion
            // ... other fields ...
```

bflad commented on website/docs/plugin/framework/functions/implementation.mdx:
```diff
@@ -232,12 +248,19 @@ func (f *ExampleFunction) Definition(ctx context.Context, req function.Definitio
             Name: "variadic_param",
-        },
+            // ... other fields ...
+		},
```
```suggestion
        },
```

bflad commented on website/docs/plugin/framework/functions/parameters/index.mdx:
````diff
@@ -46,3 +46,36 @@ Parameter type that accepts a structure of explicit attribute names.
+
+Attempting to use unnamed parameters will generate runtime errors of the following form:
+
+```shell
````
``````suggestion
```text
``````

bflad commented on website/docs/plugin/framework/functions/parameters/index.mdx:
````diff
@@ -46,3 +46,36 @@ Parameter type that accepts a structure of explicit attribute names.
+
+Parameter names are used in runtime errors to highlight which parameter is causing the issue. For example, using a value that is incompatible with the parameter type will generate an error message such as the following:
+
+```shell
````
``````suggestion
```text
``````

bflad reviewed (approved):
Looks good to me after the documentation updates, just some minor formatting nits 🚀

austinvalle reviewed (approved):
I merged #931, so I saved you the headache of the merge conflict (note 5c1f575 + 9b2199b). I had already fixed the documentation for `website/docs/plugin/framework/functions/parameters/dynamic.mdx` in my PR.

Did one last review of your PR to make sure it was all good after merge and looks clean now! 🚀

bendbennett commented:
> I merged #931, so I saved you the headache of the merge conflict (note [5c1f575](https://github.com/hashicorp/terraform-plugin-framework/commit/5c1f5750d508c0b4791a8a263cd94df54225d3b5) + [9b2199b](https://github.com/hashicorp/terraform-plugin-framework/commit/9b2199b99c6572789e46f3a9b58feb1b4ba07ab9)). I had already fixed the documentation for `website/docs/plugin/framework/functions/parameters/dynamic.mdx` in my PR.
> 
> Did one last review of your PR to make sure it was all good after merge and looks clean now from my perspective! 🚀

Thank you. Much appreciated!
